### PR TITLE
Add Support for Annex K functions according to C11

### PR DESCRIPTION
**Pull request by mostafa-salmaan**
Adding an implementation of the bounds-checking C functions (as specified in Annex K of the C11 standard) to the PicoLibc. These functions lower the risk of introducing security vulnerabilities such as buffer overflows and format string vulnerabilities into your code by providing clear and easy-to-use interfaces. For each C function a secure alternate function ending in a "_s" postfix is provided (e.g., strcpy_s). Use of these functions is recommended by security experts and secure coding standards.

### CODE REVIEW

File: newlib/libc/string/memcpy_s.c [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+#include <stdlib.h>
+
+errno_t memcpy_s(void* restrict s1, rsize_t s1max, const void* restrict s2, rsize_t n)
+{
+    bool constraint_failure = false;
+    bool zero_dest = true;
+    const char* msg = "";
+    errno_t rtn = 0;
+
+    if (s1 == NULL)
```
Comment (keith-packard) [inline newlib/libc/string/memcpy_s.c:48]: I'd find this code easier to read if the constraint failure code were below the normal return and managed with a goto at each constraint failure check point.

File: newlib/libc/string/strerror_s.c [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+    else
+    {
+        /* Standard says any value of type int shall be mapped to a message */
+        cp = "unknown error";
+    }
+    return cp;
+}
+
+/* C11 version; required by LLVM's C++11 library */
+errno_t strerror_s(char *buf, rsize_t buflen, errno_t errnum)
```
Comment (keith-packard) [inline newlib/libc/string/strerror_s.c:213]: Let's use the existing strerror mechanism (yeah, it's pretty clunky) instead of replicating the strings here.

File: test/test-memcpy_s.c [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#define __STDC_WANT_LIB_EXT1__ 1
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+
+#undef cur_handler
```
Comment (keith-packard) [inline test/test-memcpy_s.c:42]: I don't understand what this constraint handler stuff is trying to do?

File: test/test-memcpy_s.c [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+constraint_handler_t *___pcur_handler(void) {
+	return &cur_handler;
+	}
+
+int main(void)
+{
+    char src[] = "ABCDEF";
+    char dst[10];
+
+    errno_t err = memcpy_s( dst, sizeof(dst), src, strlen(src)+1 );
```
Comment (keith-packard) [inline test/test-memcpy_s.c:54]: We're going to need tests catching faults as well as successful ones to make sure the function works as specified.

**Review by keith-packard — CHANGES_REQUESTED** [review]
This looks so useful, and will be nice to have merged in once it's ready. It looks like there are a couple of changes needed to conform to the spec:

 1. Use private types in all public headers so that `stddef.h` isn't required.
 2. Fix the default value of `RSIZE_MAX`
 3. Remove duplicate definition of `set_constraint_handler_s` somehow

I'd also like to see tests that hit failure cases so that we can test those code paths as well. That should also allow testing of the constraint handler management.

Given the variable value for `RSIZE_MAX`, it would be good to have a test which checks that functionality too.

**Comment by mostafa-salmaan** [comment]
Hi @keith-packard 
Thank you for your detailed review and feedback. I have addressed all your review comments and made the requested changes. However, I am still working on the changes related to testing, particularly for testing failure cases and the functionality of the variable value of `RSIZE_MAX`. I will push these changes once they are complete.
Could you please review the current changes?

File: newlib/libc/include/errno.h [hunk 5]
```diff
@@ -36,9 +36,11 @@ SUCH DAMAGE.
 
 #ifndef __error_t_defined
 typedef int error_t;
+
```
Comment (keith-packard) [inline newlib/libc/include/errno.h:39]: Spurious white space change.

File: newlib/libc/include/errno.h [hunk 6]
```diff
@@ -36,9 +36,11 @@ SUCH DAMAGE.
...
 #define __error_t_defined 1
 #endif
 
 #include <sys/errno.h>
+#include <sys/_types.h>
```
Comment (keith-packard) [inline newlib/libc/include/errno.h:44]: This doesn't appear to be necessary in this file?

Comment (keith-packard) [reply newlib/libc/include/errno.h:44]: Ah, I see the issue. The C standard says that errno.h defines `errno_t` if `__STDC_WANT_LIB_EXT1__` is defined before the inclusion of errno.h. Further, `stdlib.h` defines the various constraint handler functions to take parameters of type `errno_t`. But, it does not say that stdlib.h defines `errno_t`, so we must not define it in that case.

Here's what we do to make that work in picolibc by creating an internal type, `__errno_t`, so that the public `errno_t` type doesn't appear except in the correct situation:

Define `__errno_t` in sys/_types.h:
```
typedef int __errno_t;
```

Include sys/_types (to get __errno_t defined) and define `errno_t` in errno.h when it should be:
```
#include <sys/_types.h>
#ifdef __STDC_WANT_LIB_EXT1__
typedef __errno_t errno_t;
#endif
```

Use `__errno_t` in sys/errno.h instead of the bare 'int' used currently. Note *do not use `errno_t`* as that may not be defined at all. 

Use `__errno_t` in stdlib.h when defining the constraint_handler functions.
```
#ifdef __STDC_WANT_LIB_EXT1__
#include <sys/_types.h>
...
void abort_handler_s(const char * restrict msg, void * restrict ptr,
__errno_t error);
...
#endif
```

The requirement here is that `errno_t` must be define if you include errno.h with `__STDC_WANT_LIB_EXT1__` defined, otherwise it must not be defined. You must also be able to include stdlib.h without having included errno.h, even if `__STDC_WANT_LIB_EXT1__` is defined, and in that case, `errno_t` must not be defined, *but* the constraint_handler API must be declared, even though you cannot use it unless you also include errno.h (as the `errno_t` type won't be defined).

File: newlib/libc/string/memcpy_s.c [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#define __STDC_WANT_LIB_EXT1__ 1
+#include <string.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
```
Comment (keith-packard) [inline newlib/libc/string/memcpy_s.c:39]: This should either use `__errno_t` or `#include <errno.h>` so that `errno_t` is defined.

**Review by keith-packard — CHANGES_REQUESTED** [review]
Fix up handling of `errno_t` vs `__errno_t`.

**Comment by mostafa-salmaan** [comment]
I have resolved all the review comments and squashed all the commits into a single one. 
@keith-packard Could you please review the changes?

**Comment by mostafa-salmaan** [comment]
> I have resolved all the review comments and squashed all the commits into a single one. @keith-packard Could you please review the changes?

I also updated the tests to cover various scenarios including normal operation, boundary conditions, and error handling.
@keith-packard Please review the changes, and let me know if there's anything needs adjustment or improvement.

**Comment by mostafa-salmaan** [comment]
During the pipeline checks, I encountered the following error message:

> ../newlib/libc/string/memcpy_s.c:52:15: error: comparison is always false due to limited range of data type [-Werror=type-limits]
> 2024-07-31T09:47:28.6816028Z    52 |     if (s1max > RSIZE_MAX)

It seems this error arises because the comparison `if (s1max > RSIZE_MAX)` is considered always false, as `RSIZE_MAX` is defined as the maximum value for `rsize_t`, which is typically the same as `size_t`.

However, according to the C17 standard, we should check that `s1max` is not greater than `RSIZE_MAX` because `RSIZE_MAX` can be configured by the user and may not necessarily equal the maximum range of the `rsize_t` type. Therefore, the check against `RSIZE_MAX` is indeed necessary to ensure that user-defined limits are respected.


I also updated all the tests to cover various scenarios including normal operation, boundary conditions, and error handling.
@keith-packard Please review the changes, and let me know if there's anything needs adjustment or improvement.

File: test/test-memset_s.c [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+#include <string.h>
+#include <errno.h>
+
+void custom_constraint_handler(const char *restrict msg, void *restrict ptr, __errno_t error);
+
+void custom_constraint_handler(const char *restrict msg, void *restrict ptr, __errno_t error)
+{
+    (void) ptr;
+    (void) error;
+    printf("Custom constraint handler called: %s\n", msg);
```
Comment (keith-packard) [inline test/test-memset_s.c:48]: You need to track that the custom handler is called at the right time and include that value in the test results.

**Review by keith-packard — CHANGES_REQUESTED** [review]
This is really starting to look great! Thanks so much for the massive improvement in the tests. I'll let you iterate until CI is happy and then take another pass over the code though.

File: newlib/libc/string/strcat_s.c [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    /* It is a constraint violation if s1max is not large enough to contain
+     * the concatenation s2: no truncation permitted.
+     * It is also a constraint violation if the string pointed to by s2
+     * overlaps s1 in any way.
+     * The C11 Rationale says we are permitted to proceed with the copy and
+     * detect dest buffer overrun and overlapping memory blocks as a byproduct
+     * of performing the copy operation.  This is to avoid calling strlen on
+     * s2 to detect these violations prior to attempting the copy.
+     */
+	// compute chars available in s1
```
Comment (arichardson) [inline newlib/libc/string/strcat_s.c:77]: formatting is a bit weird here, maybe just run the whole thing through `git clang-format`?

Comment (mostafa-salmaan) [reply newlib/libc/string/strcat_s.c]: Thank you for pointing that out. I've applied `git clang-format` to address the formatting issues and pushed the changes.

File: test/test-memset_s.c (hunk 8 above)
Comment (mostafa-salmaan) [reply test/test-memset_s.c]: > You need to track that the custom handler is called at the right time and include that value in the test results.

I've updated all the tests to track that the custom handler is called at the right time and include that value in the test results.

**Comment by mostafa-salmaan** [comment]
> This is really starting to look great! Thanks so much for the massive improvement in the tests. I'll let you iterate until CI is happy and then take another pass over the code though.

@keith-packard Could you please check my comment here https://github.com/picolibc/picolibc/pull/744#issuecomment-2260295254
I noticed that the CI checks have one repeated fail in `memmove_s`, `memcpy_s`, `memset_s`, and `strcat_s`. I believe this is not an issue as described in my comment.

**Comment by keith-packard** [comment]
> @keith-packard Could you please check my comment here [#744 (comment)](https://github.com/picolibc/picolibc/pull/744#issuecomment-2260295254) I noticed that the CI checks have one repeated fail in `memmove_s`, `memcpy_s`, `memset_s`, and `strcat_s`. I believe this is not an issue as described in my comment.

You're correct; we don't want these compiler warnings to be generated. K.3.4 mentions this as a possibility:

> Implementations targeting machines with small address spaces may wish to define RSIZE_MAX as SIZE_MAX, which
means that there is no object size that is considered a runtime-constraint violation.

Because of this, we need to elide the size checks entirely when `RSIZE_MAX == SIZE_MAX`.

I'd suggest creating a new internal header file, `newlib/libc/stdlib/local_s.h` which you can use to hold common macros to use while implementing this Annex. You could place something like this:
```
#if RSIZE_MAX == SIZE_MAX
#define CHECK_RSIZE(s)  0
#else
#define CHECK_RSIZE(s)  (RSIZE_MAX == SIZE_MAX ? 0 : (s) > RSIZE_MAX)
#endif
```
into that file and then use this macro when checking `rsize_t` values.

I also noticed a pattern you're using in the code to fetch the current constraint handler value:
```
    handler = set_constraint_handler_s(NULL);
    (void)set_constraint_handler_s(handler);
```
We can't do that -- the constraint handler is a global value shared by all threads; this code changes the value between the two function calls. Instead, declare your `__cur_handler` variable in `local_s.h` and then reference it directly where used.

**Review by keith-packard — CHANGES_REQUESTED** [review]
Clean up the compiler warnings by skipping the `RSIZE_MAX` checks when `RSIZE_MAX == SIZE_MAX`.
Use `__cur_handler` directly instead of `set_constraint_handler_s` to avoid leaving the constraint handler unset for other threads.

File: newlib/libc/include/stdint.h [hunk 10]
```diff
@@ -457,6 +457,17 @@ typedef __uint_least64_t uint_least64_t;
...
+#if __STDC_WANT_LIB_EXT1__ == 1
+#include <sys/_types.h>
 
-#endif /* _STDINT_H */
+  // could be defined by the user
+#ifndef RSIZE_MAX
+#define RSIZE_MAX SIZE_MAX
+#endif
+
+  typedef __rsize_t rsize_t;
```
Comment (keith-packard) [inline newlib/libc/include/stdint.h:468]: Looks like this is supposed to be in stddef.h instead of stdint.h? You should be able to remove the include of <sys/_types.h> from this file as that won't be needed.

File: newlib/libc/include/errno.h [hunk 11]
```diff
@@ -43,4 +43,9 @@ typedef int error_t;
 
 #include <sys/errno.h>
 
+#include <sys/_types.h>
+#ifdef __STDC_WANT_LIB_EXT1__
+typedef __errno_t errno_t;
```
Comment (keith-packard) [inline newlib/libc/include/errno.h:48]: This also needs to be in stdlib.h -- we'll have to use the same pattern as other typedefs which are required from multiple files:
```
#ifndef _ERRNO_T_DEFINED
typedef __errno_t errno_t;
#define _ERRNO_T_DEFINED
#endif
```
Stick that in errno.h, stdio.h (both tinystdio and legacy stdio), stdlib.h, time.h, string.h, and wchar.h.

File: newlib/libc/include/stdint.h (hunk 10 above)
Comment (keith-packard) [inline newlib/libc/include/stdint.h:468]: Like `errno_t`, `rsize_t` must be defined from multiple files: stddef.h, stdio.h, stdlib.h, string.h, time.h and wchar.h. Use the same pattern as `errno_t`:
```
#ifndef _RSIZE_T_DEFINED
typedef __rsize_t rsize_t;
#define _RSIZE_T_DEFINED
#endif
```

File: newlib/libc/include/stdlib.h [hunk 12]
```diff
@@ -408,10 +408,22 @@ char   *__ldtoa (long double, int, int, int *, int *, char **);
 void	__eprintf (const char *, const char *, unsigned int, const char *);
 #endif
 
+#if __STDC_WANT_LIB_EXT1__ == 1
+#include <sys/_types.h>
+
+typedef void (*constraint_handler_t)(const char *restrict msg,
+                                     void *restrict ptr, __errno_t error);
+
+extern constraint_handler_t
```
Comment (keith-packard) [inline newlib/libc/include/stdlib.h:417]: I'd prefer you leave 'extern' off function declarations; it's unnecessary.

File: newlib/libc/include/string.h [hunk 13]
```diff
@@ -201,6 +201,26 @@ int	 timingsafe_bcmp (const void *, const void *, size_t);
 int	 timingsafe_memcmp (const void *, const void *, size_t);
 #endif
 
+#if __STDC_WANT_LIB_EXT1__ == 1
+#include <sys/_types.h>
+
+typedef __rsize_t rsize_t;
+
+extern __errno_t memcpy_s(void *__restrict, rsize_t, const void *__restrict,
```
Comment (keith-packard) [inline newlib/libc/include/string.h:209]: Please remove `extern`

File: newlib/libc/include/sys/_types.h [hunk 14]
```diff
@@ -258,4 +258,21 @@ typedef	unsigned short	__nlink_t;
...
 typedef	unsigned long	__useconds_t;	/* microseconds (unsigned) */
 
+#ifdef __STDC_WANT_LIB_EXT1__
+#if (__STDC_WANT_LIB_EXT1__ != 0) && (__STDC_WANT_LIB_EXT1__ != 1)
+#error Please define __STDC_WANT_LIB_EXT__ as 0 or 1
+#endif
+
+#if __STDC_WANT_LIB_EXT1__ == 1
+
+#ifndef __RSIZE_T
```
Comment (keith-packard) [inline newlib/libc/include/sys/_types.h:268]: We don't need to protect the __rsize_t typedef; it can be unconditional, just like __errno_t.

File: newlib/libc/include/sys/errno.h [hunk 15]
```diff
@@ -35,6 +35,7 @@ SUCH DAMAGE.
 
 #include <sys/cdefs.h>
 #include <sys/config.h>
+#include <sys/_types.h>
```
Comment (keith-packard) [inline newlib/libc/include/sys/errno.h:38]: This looks like it's left over from a previous edit?

File: test/test-memcpy_s.c [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+
+void custom_constraint_handler(const char *restrict msg, void *restrict ptr,
+                               __errno_t error);
+int test_handler_called(int handler_called, char *expected_msg, int test_id);
+
+char handler_msg[MAX_ERROR_MSG] = "";
+
+void
+custom_constraint_handler(const char *restrict msg, void *restrict ptr,
+                          __errno_t error)
```
Comment (keith-packard) [inline test/test-memcpy_s.c:52]: Tests should use the public API, so replace `__errno_t` with `errno_t`.

File: test/test-memcpy_s.c [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+
+#define __STDC_WANT_LIB_EXT1__ 1
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+
+#define MAX_ERROR_MSG 100
+
+void custom_constraint_handler(const char *restrict msg, void *restrict ptr,
```
Comment (keith-packard) [inline test/test-memcpy_s.c:44]: If `custom_constraint_handler` were declared static, you wouldn't need this forward declaration.

**Review by keith-packard — CHANGES_REQUESTED** [review]
The tests look great; all of the library code is in good shape. Once the header files are cleaned up to follow the POSIX/C rules, this series will be ready to merge. Thanks so much!

**Comment by keith-packard** [comment]
This is coming along nicely. Can I get you to review the Annex K spec and make sure `errno_t` and `rsize_t` are defined in all of the required include files? These all need to use the pattern described above so that the types are not defined more than once when multiple include files are used.

`errno_t`: errno.h, stdio.h, stdlib.h, string.h, time.h, wchar.h
`rsize_t`: stddef.h, stdio.h, stdlib.h, string.h, time.h, wchar.h

**Comment by mostafa-salmaan** [comment]
Now, `errno_t` and `rsize_t` are defined in all of the required include files with the pattern described above.
@keith-packard Could you please review the changes and confirm if this series is ready to be merged?

**Comment by mostafa-salmaan** [comment]
@keith-packard Could you please let me know if there are any pending actions required on our side for this PR? 
If everything is ready for merge, could you provide details on what might be pending or any remaining steps?

**Comment by keith-packard** [comment]
Sorry, I'm on holiday and not keeping up with email very effectively. I've just pushed a formatting update as a few files were missing final newlines. You don't need to do anything further here; once CI is happy, I'll be merging this in. Thanks again for your huge contribution to picolibc!

**Comment by mostafa-salmaan** [comment]
> Sorry, I'm on holiday and not keeping up with email very effectively. I've just pushed a formatting update as a few files were missing final newlines. You don't need to do anything further here; once CI is happy, I'll be merging this in. Thanks again for your huge contribution to picolibc!

Thank you, Keith! I noticed the PR has been merged. I appreciate your effort in getting this finalized.
